microhapulator/qcstats.py

```diff
@@ -0,0 +1,139 @@
+# -------------------------------------------------------------------------------------------------
+# Copyright (c) 2024, DHS.
+#
+# This file is part of MicroHapulator (https://github.com/bioforensics/microhapulator) and is
+# licensed under the BSD license: see LICENSE.txt.
+#
+# This software was prepared for the Department of Homeland Security (DHS) by the Battelle National
+# Biodefense Institute, LLC (BNBI) as part of contract HSHQDC-15-C-00064 to manage and operate the
+# National Biodefense Analysis and Countermeasures Center (NBACC), a Federally Funded Research and
+# Development Center.
+# -------------------------------------------------------------------------------------------------
+
+from dataclasses import dataclass
+import pandas as pd
+import re
+
+
+@dataclass
+class FilterStats:
+    filtered: int
+    retained: int
+
+    @property
+    def total(self):
+        return self.filtered + self.retained
+
+    @property
+    def total_reads(self):
+        return f"{self.total:,}"
+
+    @property
+    def excluded(self):
+        return f"{self.filtered:,}"
+
+    @property
+    def kept(self):
+        return f"{self.retained:,}"
+
+    @property
+    def retention_rate(self):
+        rate = self.retained / self.total * 100
+        return f"{rate:.1f}%"
+
+    @classmethod
+    def from_txt(cls, filepath):
+        table = pd.read_csv(filepath, sep="\t")
+        assert len(table) == 1
+        entry = table.iloc[0]
+        stats = cls(entry.ReadsRemoved, entry.ReadsKept)
+        return stats
+
+
+@dataclass
+class PairedAmbiguityFilterStats:
+    total: int
+    failed_r1: int
+    failed_r2: int
+    failed_both: int
+    retained: int
```

Comment:
Ambiguity filtering stats, keeping track of which fragment(s) are responsible for the pair being discarded.